rama-http/src/service/web/router.rs

```diff
@@ -0,0 +1,442 @@
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::POST);
+        self.add_route(path, matcher, service)
+    }
+
+    /// add a PUT route to the router.
+    pub fn put<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::PUT);
+        self.add_route(path, matcher, service)
+    }
+
+    /// add a DELETE route to the router.
+    pub fn delete<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::DELETE);
+        self.add_route(path, matcher, service)
+    }
+
+    /// add a PATCH route to the router.
+    pub fn patch<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::PATCH);
+        self.add_route(path, matcher, service)
+    }
+
+    /// add a HEAD route to the router.
+    pub fn head<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::HEAD);
+        self.add_route(path, matcher, service)
+    }
+
+    /// add a OPTIONS route to the router.
+    pub fn options<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::OPTIONS);
+        self.add_route(path, matcher, service)
+    }
+
+    /// add a TRACE route to the router.
+    pub fn trace<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::TRACE);
+        self.add_route(path, matcher, service)
+    }
+
+    /// add a CONNECT route to the router.
+    pub fn connect<I, T>(self, path: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let matcher = HttpMatcher::method(MethodMatcher::CONNECT);
+        self.add_route(path, matcher, service)
+    }
+
+    /// register a nested router under a prefix.
+    ///
+    /// The prefix is used to match the request path and strip it from the request URI.
+    pub fn sub<I, T>(self, prefix: &str, service: I) -> Self
+    where
+        I: IntoEndpointService<State, T>,
+    {
+        let path = format!("{}/{}", prefix.trim_end_matches(['/']), "{*nest}");
+        let matcher = HttpMatcher::custom(true);
+
+        let nested_router_service = NestedRouterService {
+            prefix: prefix.to_owned(),
+            nested: service.into_endpoint_service().boxed(),
+        };
+
+        self.add_route(&path, matcher, nested_router_service)
+    }
+
+    /// add a route to the router with it's matcher and service.
+    fn add_route<I, T>(mut self, path: &str, matcher: HttpMatcher<State, Body>, service: I) -> Self
```

Comment:
can you rename this to `match` and make it public. This allows one to use a custom matcher :)